AADServicePrincipal - BUG - Subsequent run Fails on App Display name

### Description of the issue

DSC is able to provision and App Registration (AADApplication) and Service principal (AADServicePrincipal) with the AADServicePrincipal.AppId = AADApplication.DisplayName. However, a subsequent run of DSC fails for AAD Service Principal. 

![Image](https://github.com/user-attachments/assets/64f92750-f11f-4f4b-96cd-beb3160d391a)


Updating the AADServicePrincipal.AppId = AADApplication.ObjectId corrects the issue.


Related Bug: https://github.com/microsoft/Microsoft365DSC/issues/4965#issuecomment-2307282915

### Microsoft 365 DSC Version

1.24.1016.1

### Which workloads are affected

Azure Active Directory (Entra ID)

### The DSC configuration

```
AADApplication 'AADApplication-NEWTESTAPP' {
    DependsOn               = @(
        '[AADApplication]AADApplication-Microsoft365DSC'
    )
    AvailableToOtherTenants = $False
    DisplayName             = 'NEWTESTAPP'
    Description             = 'NEWTESTAPP'
    Ensure                  = 'Present'
    # IdentifierUris          = @()
    KnownClientApplications = @()
    IsFallbackPublicClient  = $False
    Owners                  = @(
        $AppOwners.IliadIdentityApp
    )
    PublicClient            = $False
    ReplyURLs               = @()
    ApplicationId           = $ApplicationId
    TenantId                = $TenantId
    CertificateThumbprint   = $Thumbprint
}

# https://microsoft365dsc.com/resources/azure-ad/AADServicePrincipal/
AADServicePrincipal 'AADServicePrincipal-NEWTESTAPPSP' {
    DependsOn               = @(
        '[AADApplication]AADApplication-Microsoft365DSC'
    )
    AppId = 'NEWTESTAPP'
    DisplayName             = 'NEWTESTAPP'
    Ensure                  = 'Present'
    ServicePrincipalType    = 'Application'
    ApplicationId           = $ApplicationId
    TenantId                = $TenantId
    CertificateThumbprint   = $Thumbprint
}
```

### Verbose logs showing the problem
```
##[error][Request_MultipleObjectsWithSameKeyValue] : The service principal cannot be created, updated, or restored because the 
service principal name c83defb3-cabc-48a2-b697-2a8a218e6a6f is already in use.
    + CategoryInfo          : InvalidOperation: ({ Headers = , b...vicePrincipal }:) [], CimException
    + FullyQualifiedErrorId : Request_MultipleObjectsWithSameKeyValue,Microsoft.Graph.PowerShell.Cmdlets.NewMgServiceP 
   rincipal_CreateExpanded
```

### Environment Information + PowerShell Version

```
OsName               : Microsoft Windows Server 2022 Datacenter
OsOperatingSystemSKU : DatacenterServerEdition
OsArchitecture       : 64-bit
WindowsVersion       : 2009
WindowsBuildLabEx    : 20348.1.amd64fre.fe_release.210507-1500
OsLanguage           : en-US
OsMuiLanguages       : {en-US}
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -29,6 +29,9 @@
   * Initial release.
 * AADFilteringProfile
   * Initial release.
+* AADGroup
+  * Added support for custom roles assignment.
+    FIXES [#5322](https://github.com/microsoft/Microsoft365DSC/issues/5322)
 * AADHomeRealmDiscoveryPolicy
   * Initial Release
 * AADIdentityAPIConnector

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -64,6 +64,7 @@
   * Initial release.
 * AADServicePrincipal
   * Added the notes field.
+  * FIXES [#5312](https://github.com/microsoft/Microsoft365DSC/issues/5312)
 * AADSocialIdentityProvider
   * Fixed missing permissions in settings.json
 * AADVerifiedIdAuthority

--- a/Modules/Microsoft365DSC/DSCResources/MSFT_AADGroup/MSFT_AADGroup.psm1
+++ b/Modules/Microsoft365DSC/DSCResources/MSFT_AADGroup/MSFT_AADGroup.psm1
@@ -256,20 +256,17 @@ function Get-TargetResource
             if ($Group.IsAssignableToRole -eq $true)
             {
                 $AssignedToRoleValues = @()
-                # Note: only process directory roles and not group membership (if any)
-                foreach ($role in $($memberOf | Where-Object -FilterScript { $_.AdditionalProperties.'@odata.type' -eq '#microsoft.graph.directoryRole' }))
+                $roleAssignments = Get-MgBetaRoleManagementDirectoryRoleAssignment -Filter "PrincipalId eq '$($Group.Id)'"
+                foreach ($assignment in $roleAssignments)
                 {
-                    if ($null -ne $role.AdditionalProperties.displayName)
-                    {
-                        $AssignedToRoleValues += $role.AdditionalProperties.displayName
-                    }
+                    $roleDefinition = Get-MgBetaRoleManagementDirectoryRoleDefinition -UnifiedRoleDefinitionId $assignment.RoleDefinitionId
+                    $AssignedToRoleValues += $roleDefinition.DisplayName
                 }
             }
 
             # Licenses
             $assignedLicensesValues = $null
             $uri = $Global:MSCloudLoginConnectionProfile.MicrosoftGraph.ResourceUrl + "v1.0/groups/$($Group.Id)/assignedLicenses"
-            
             $assignedLicensesRequest = Invoke-MgGraphRequest -Method 'GET' `
                 -Uri $uri
 
@@ -914,13 +911,7 @@ function Set-TargetResource
             {
                 try
                 {
-                    $role = Get-MgBetaDirectoryRole -Filter "DisplayName eq '$($diff.InputObject)'"
-                    # If the role hasn't been activated, we need to get the role template ID to first activate the role
-                    if ($null -eq $role)
-                    {
-                        $adminRoleTemplate = Get-MgBetaDirectoryRoleTemplate -All | Where-Object { $_.DisplayName -eq $diff.InputObject }
-                        $role = New-MgBetaDirectoryRole -RoleTemplateId $adminRoleTemplate.Id
-                    }
+                    $role = Get-MgBetaRoleManagementDirectoryRoleDefinition -Filter "DisplayName eq '$($diff.InputObject)'"
                 }
                 catch
                 {
@@ -935,15 +926,15 @@ function Set-TargetResource
                     if ($diff.SideIndicator -eq '=>')
                     {
                         Write-Verbose -Message "Assigning AAD group {$($currentGroup.DisplayName)} to Directory Role {$($diff.InputObject)}"
-                        $DirObject = @{
-                            '@odata.id' = "https://graph.microsoft.com/v1.0/directoryObjects/$($currentGroup.Id)"
-                        }
-                        New-MgBetaDirectoryRoleMemberByRef -DirectoryRoleId ($role.Id) -BodyParameter $DirObject | Out-Null
+                        New-MgBetaRoleManagementDirectoryRoleAssignment -RoleDefinitionId $role.Id -PrincipalId $currentGroup.Id -DirectoryScopeId '/'
                     }
                     elseif ($diff.SideIndicator -eq '<=')
                     {
                         Write-Verbose -Message "Removing AAD group {$($currentGroup.DisplayName)} from Directory Role {$($role.DisplayName)}"
-                        Remove-MgBetaDirectoryRoleMemberDirectoryObjectByRef -DirectoryRoleId ($role.Id) -DirectoryObjectId ($currentGroup.Id) | Out-Null
+                        Write-Verbose "GroupId = $($currentGroup.Id)"
+                        Write-Verbose "RoleDefinitionId = $($role.Id)"
+                        $roleAssignment = Get-MgBetaRoleManagementDirectoryRoleAssignment -Filter "PrincipalId eq '$($currentGroup.Id)' and RoleDefinitionId eq '$($role.Id)'"
+                        Remove-MgBetaRoleManagementDirectoryRoleAssignment -UnifiedRoleAssignmentId $roleAssignment.Id
                     }
                 }
             }

--- a/Modules/Microsoft365DSC/DSCResources/MSFT_AADServicePrincipal/MSFT_AADServicePrincipal.psm1
+++ b/Modules/Microsoft365DSC/DSCResources/MSFT_AADServicePrincipal/MSFT_AADServicePrincipal.psm1
@@ -237,7 +237,7 @@ function Get-TargetResource
             }
 
             [Array]$complexDelegatedPermissionClassifications = @()
-            $Uri = $Global:MSCloudLoginConnectionProfile.MicrosoftGraph.ResourceUrl + "v1.0/servicePrincipals(appId='$AppId')/delegatedPermissionClassifications"
+            $Uri = $Global:MSCloudLoginConnectionProfile.MicrosoftGraph.ResourceUrl + "v1.0/servicePrincipals/$($AADServicePrincipal.Id)/delegatedPermissionClassifications"
             $permissionClassifications = Invoke-MgGraphRequest -Uri $Uri -Method Get
             foreach ($permissionClassification in $permissionClassifications.Value){
                 $hashtable = @{
@@ -301,13 +301,13 @@ function Get-TargetResource
                 }
             }
 
-            $complexCustomSecurityAttributes = [Array](Get-CustomSecurityAttributes -AppId $AppId)
+            $complexCustomSecurityAttributes = [Array](Get-CustomSecurityAttributes -ServicePrincipalId $AADServicePrincipal.Id)
             if ($null -eq $complexCustomSecurityAttributes) {
                 $complexCustomSecurityAttributes = @()
             }
 
             $result = @{
-                AppId                              = $AADServicePrincipal.AppId
+                AppId                              = $appInstance.DisplayName
                 AppRoleAssignedTo                  = $AppRoleAssignedToValues
                 ObjectID                           = $AADServicePrincipal.Id
                 DisplayName                        = $AADServicePrincipal.DisplayName
@@ -1220,10 +1220,10 @@ function Create-AttributeValue {
 function Get-CustomSecurityAttributes {
     [OutputType([System.Array])]
     param (
-        [String]$AppId
+        [String]$ServicePrincipalId
     )
 
-    $customSecurityAttributes = Invoke-MgGraphRequest -Uri "https://graph.microsoft.com/beta/servicePrincipals(appId='$AppId')`?`$select=customSecurityAttributes" -Method Get
+    $customSecurityAttributes = Invoke-MgGraphRequest -Uri "https://graph.microsoft.com/beta/servicePrincipals/$($ServicePrincipalId)`?`$select=customSecurityAttributes" -Method Get
     $customSecurityAttributes = $customSecurityAttributes.customSecurityAttributes
     $newCustomSecurityAttributes = @()
 

--- a/Tests/Unit/Microsoft365DSC/Microsoft365DSC.AADGroup.Tests.ps1
+++ b/Tests/Unit/Microsoft365DSC/Microsoft365DSC.AADGroup.Tests.ps1
@@ -81,13 +81,16 @@ Describe -Name $Global:DscHelper.DescribeHeader -Fixture {
             Mock -CommandName New-MgBetaGroupOwnerByRef -MockWith {
             }
 
-            Mock -CommandName New-MgBetaGroupMemberByRef -MockWith {
+            Mock -CommandName Remove-MgBetaRoleManagementDirectoryRoleAssignment -MockWith {
             }
 
             Mock -CommandName New-MgBetaGroupMember -MockWith {
             }
 
             Mock -CommandName New-MgBetaDirectoryRoleMemberByRef -MockWith {
+            }
+
+            Mock -CommandName New-MgBetaRoleManagementDirectoryRoleAssignment -MockWith {
             }
 
             Mock -CommandName Remove-MgGroupOwnerDirectoryObjectByRef -MockWith {
@@ -103,6 +106,18 @@ Describe -Name $Global:DscHelper.DescribeHeader -Fixture {
             }
 
             Mock -CommandName Set-MgGroupLicense -MockWith {
+            }
+
+            Mock -CommandName Get-MgBetaRoleManagementDirectoryRoleAssignment -MockWith {
+                return @{
+                    RoleDefinitionId = "12345-12345-12345-12345-12345"
+                }
+            }
+
+            Mock -CommandName Get-MgBetaRoleManagementDirectoryRoleDefinition -MockWith {
+                return @{
+                    DisplayName = "AADRole"
+                }
             }
 
             # Mock Write-Host to hide output during the tests
@@ -508,6 +523,9 @@ Describe -Name $Global:DscHelper.DescribeHeader -Fixture {
                         ID          = '12345-12345-12345-12345'
                     }
                 }
+                Mock -CommandName Get-MgBetaRoleManagementDirectoryRoleAssignment -MockWith {
+                    return $null
+                }
             }
 
             It 'Should return Values from the Get method' {
@@ -522,8 +540,8 @@ Describe -Name $Global:DscHelper.DescribeHeader -Fixture {
             It 'Should call the Set method' {
                 Set-TargetResource @testParams
                 Should -Invoke -CommandName 'Get-MgGroup' -Exactly 1
-                Should -Invoke -CommandName 'Get-MgBetaDirectoryRole' -Exactly 1
-                Should -Invoke -CommandName 'New-MgBetaDirectoryRoleMemberByRef' -Exactly 1
+                Should -Invoke -CommandName 'Get-MgBetaRoleManagementDirectoryRoleAssignment' -Exactly 1
+                Should -Invoke -CommandName 'New-MgBetaRoleManagementDirectoryRoleAssignment' -Exactly 1
             }
         }
 
@@ -646,8 +664,8 @@ Describe -Name $Global:DscHelper.DescribeHeader -Fixture {
             It 'Should call the Set method' {
                 Set-TargetResource @testParams
                 Should -Invoke -CommandName 'Get-MgGroup' -Exactly 1
-                Should -Invoke -CommandName 'Get-MgBetaDirectoryRole' -Exactly 1
-                Should -Invoke -CommandName 'Remove-MgBetaDirectoryRoleMemberDirectoryObjectByRef' -Exactly 1
+                Should -Invoke -CommandName 'Get-MgBetaRoleManagementDirectoryRoleDefinition' -Exactly 2
+                Should -Invoke -CommandName 'Remove-MgBetaRoleManagementDirectoryRoleAssignment' -Exactly 1
             }
         }
 

--- a/Tests/Unit/Microsoft365DSC/Microsoft365DSC.AADServicePrincipal.Tests.ps1
+++ b/Tests/Unit/Microsoft365DSC/Microsoft365DSC.AADServicePrincipal.Tests.ps1
@@ -43,6 +43,13 @@ Describe -Name $Global:DscHelper.DescribeHeader -Fixture {
                 return 'Credentials'
             }
 
+            Mock -CommandName Get-MgApplication -MockWith {
+                return @{
+                    AppId = "12345-12345-12345-12345-12345"
+                    DisplayName = "App1"
+                }
+            }
+
             # Mock Write-Host to hide output during the tests
             Mock -CommandName Write-Host -MockWith {
             }
@@ -206,7 +213,7 @@ Describe -Name $Global:DscHelper.DescribeHeader -Fixture {
         Context -Name 'The app exists and values are already in the desired state' -Fixture {
             BeforeAll {
                 $testParams = @{
-                    AppId                     = 'b4f08c68-7276-4cb8-b9ae-e75fca5ff834'
+                    AppId                     = 'App1'
                     DisplayName               = 'App1'
                     AlternativeNames          = 'AlternativeName1', 'AlternativeName2'
                     AccountEnabled            = $true

